### PR TITLE
Adds annotations to secure sensible links against CSRF attacks

### PR DESCRIPTION
**Pull request by Tobsucht**

### CODE REVIEW

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 1]
```diff
@@ -168,6 +182,18 @@ private void performRouteInOwnThread(WebContext ctx, Route route, List<Object> p
         ctx.enableTiming(route.toString());
     }
 
+    private boolean checkCSRFTokenIfNecessary(WebContext ctx, Route route) {
+        if (route.getMethod().isAnnotationPresent(CheckSecurityToken.class) && (!ctx.isPOST()
```
Comment (andyHa) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: split into 3 if and return true immediatelly

Comment (andyHa) [reply src/main/java/sirius/web/controller/ControllerDispatcher.java]: *2

File: src/main/java/sirius/web/http/CSRFHelper.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * Returns the CSRF security-token of the current session. Internally recomputes the token if outdated.
+     *
+     * @return the CSRF security-token to protect sensitive links.
+     */
+    public String getCSRFToken(WebContext ctx) {
+        Value lastCSRFRecompute = ctx.getSessionValue(LAST_CSRF_RECOMPUTE);
+
+        if (isCSRFTokenOutdated(lastCSRFRecompute.asLong(-1L)) || lastCSRFRecompute.isEmptyString()) {
```
Comment (andyHa) [inline src/main/java/sirius/web/http/CSRFHelper.java]: or is redundant

File: src/main/java/sirius/web/http/WebContext.java [hunk 3]
```diff
@@ -1078,6 +1083,15 @@ public boolean isSSL() {
         return ssl;
     }
 
+    /**
+     * Returns the CSRF security-token of the current session.
+     *
+     * @return the CSRF security-token to protect sensitve links.
+     */
+    public String getCSRFToken() {
+        return csrfHelper.getCSRFToken(this);
```
Comment (andyHa) [inline src/main/java/sirius/web/http/WebContext.java]: can be left out(?)

File: src/main/resources/default/taglib/w/deleteLink.html.pasta [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
-    @i18n("NLS.delete")
-</a>
+<i:local name="deleteFormId" value="@apply('deleteForm-%s', call.generateLocalId())"/>
+
+<form id="@deleteFormId" method="post" action="@if(page != null) {@url?@page.createQueryString()} else {@url}">
+    <i class="fa fa-trash"></i>
+    <a class="link link-danger @if (adminOnly) {admin-link}" data-deleteid="@deleteFormId" href="#">
+        @i18n("NLS.delete")
+    </a>
+    <input name="CSRFToken" value="@call.getCSRFToken()" type="hidden"/>
```
Comment (andyHa) [inline src/main/resources/default/taglib/w/deleteLink.html.pasta]: add CSRFHelper to DefaultGlobalContextExtender as "csrf" and call csrf.getToken(call) here

File: src/test/java/sirius/web/controller/TestController.java [hunk 5]
```diff
@@ -24,6 +25,10 @@
 
 @Register
 public class TestController implements Controller {
+
+    @ConfigValue("http.serverSessionParameterName")
+    private static String serverSessionParameterName;
```
Comment (andyHa) [inline src/test/java/sirius/web/controller/TestController.java]: kill

File: src/test/resources/templates/security-token.html.pasta [hunk 6]
```diff
@@ -0,0 +1 @@
+@call.getCSRFToken()
```
Comment (andyHa) [inline src/test/resources/templates/security-token.html.pasta]: see above

**Comment by sabieber** [comment]
build

**Comment by scireumskip** [comment]
Can one of the admins verify this patch?

**Comment by sabieber** [comment]
build

**Comment by andyHa** [comment]
BUILD

**Comment by sabieber** [comment]
ok to test

**Comment by sabieber** [comment]
ok to test

**Comment by sabieber** [comment]
add to whitelist

**Comment by sabieber** [comment]
build

**Comment by sabieber** [comment]
build

File: src/main/resources/default/templates/wondergem/page-script.html.pasta [hunk 7]
```diff
@@ -3,13 +3,12 @@
...
             try {
-                var clickedLink = event.currentTarget;
                 var modalElement = $("#danger-link-confirm");
-
-                var okayLink = $(clickedLink).attr('href');
                 var submitBtn = modalElement.find(":submit");
+                var that = $(this);
+
                 submitBtn.click(function () {
-                    window.location.href = okayLink;
```
Comment (sabieber) [inline src/main/resources/default/templates/wondergem/page-script.html.pasta]: Please keep the old functionality and add a new listener with an extra class so old usages of link-danger dont have to be migrated right away

File: src/main/resources/default/templates/wondergem/page-script.html.pasta [hunk 8]
```diff
@@ -1,7 +1,34 @@
 <script type="text/javascript">
-
     function createConfirmForDangerousLinks(message) {
-        $("a.link-danger, a.button-danger-ask").on('click', function (event) {
+        $("a.link-danger:not(.guarded-link), a.button-danger-ask:not(.guarded-link)").on('click', function (event) {
```
Comment (andyHa) [inline src/main/resources/default/templates/wondergem/page-script.html.pasta]: wouldn't an if ($(this).hasClass('guarded-Link')) { x} else {x} be easier?

File: src/main/resources/default/templates/wondergem/page-script.html.pasta [hunk 9]
```diff
@@ -1,7 +1,34 @@
 <script type="text/javascript">
-
     function createConfirmForDangerousLinks(message) {
-        $("a.link-danger, a.button-danger-ask").on('click', function (event) {
```
Comment (andyHa) [inline src/main/resources/default/templates/wondergem/page-script.html.pasta]: the "message" parameter (above) isn't needed (?)

**Pull request merged**